src/k4a_ros_device.cpp

```diff
@@ -663,14 +752,55 @@ void K4AROSDevice::framePublisherThread()
 
     while (running_ && ros::ok() && !ros::isShuttingDown())
     {
-        // TODO: consider appropriate capture timeout based on camera framerate
-        bool success = k4a_device_.get_capture(&capture, std::chrono::milliseconds(K4A_WAIT_INFINITE));
+        bool success = true;
 
-        if (!success)
+        if (k4a_device_)
         {
-            ROS_FATAL("Failed to poll cameras: node cannot continue.");
-            ros::requestShutdown();
-            return;
+            // TODO: consider appropriate capture timeout based on camera framerate
+            if (!k4a_device_.get_capture(&capture, std::chrono::milliseconds(K4A_WAIT_INFINITE)))
+            {
+                ROS_FATAL("Failed to poll cameras: node cannot continue.");
+                ros::requestShutdown();
+                return;
+            }
+        }
+        else if (k4a_playback_handle_)
+        {
+            std::lock_guard<std::mutex> guard(k4a_playback_handle_mutex_);
+            k4a_capture_t capture_t;
+            k4a_stream_result_t stream_result = k4a_playback_get_next_capture(k4a_playback_handle_, &capture_t);
+
+            if (stream_result == K4A_STREAM_RESULT_EOF)
+            {
+                // rewind recording if looping is enabled
+                if (params_.recording_loop_enabled)
+                {
+                    if (k4a_playback_seek_timestamp(k4a_playback_handle_, 0, K4A_PLAYBACK_SEEK_BEGIN) != K4A_RESULT_SUCCEEDED)
+                    {
+                        ROS_FATAL("Error seeking recording to beginning. node cannot continue.");
+                        ros::requestShutdown();
+                        return;
+                    }
+                    stream_result = k4a_playback_get_next_capture(k4a_playback_handle_, &capture_t);
+                    imu_stream_end_of_file_ = false;
+                    last_imu_time_usec_ = 0;
+                }
+                else
+                {
+                    ROS_WARN("Recording reached end of file. node cannot continue.");
```

Comment:
Sure I can use ROS_INFO as well. I only used it so that it stood out because there is a lot of other console output when the node shuts down and I thought the user may overlook it and is confused why the node shuts down. In my case it was just a 5 sec recording and I first thought it crashed until I realised that the Recording itself was just so short that I did not open RViz fast enough :)